File: lib/core/global/responses/rest_response.dart
import 'package:sertton/core/constants/http_status_code.dart';

class RestResponse<Body> {
  final Body? _body;
  final int _statusCode;
  final String? _errorMessage;

  RestResponse({Body? body, int? statusCode, String? errorMessage})
    : _body = body,
      _statusCode = statusCode ?? HttpStatusCode.ok,
      _errorMessage = errorMessage;

  Body get body {
    if (_errorMessage != null) {
      throw Exception('Rest Response failed');
    }
    return _body!;
  }

  Never throwError() {
    throw Exception('Rest Response failed');
  }

  RestResponse<NewBody> mapBody<NewBody>(NewBody Function(Body?) mapper) {
    if (_errorMessage != null) {
      throw Exception('Rest Response failed');
    }
    return RestResponse(body: mapper(_body), statusCode: _statusCode);
  }

  bool get isSuccessful {
    return _statusCode <= HttpStatusCode.badRequest;
  }

  bool get isFailure {
    return _statusCode >= HttpStatusCode.badRequest || _errorMessage != null;
  }

  String get errorMessage {
    if (_errorMessage != null) {
      return _errorMessage!;
    }
    throw Exception('Rest Response failed');
  }

  int get statusCode => _statusCode;
}

File: src/core/shared/responses/rest-response.ts
import { HTTP_STATUS_CODE } from "@/core/shared/constants/http-status-code"
import { AppError } from "@/core/shared/errors/app-error"

export interface RestResponseProps<Body> {
  readonly body?: Body
  readonly error?: AppError
  readonly statusCode?: number
}

export class RestResponse<Body> {
  public readonly _body?: Body
  public readonly error?: AppError
  public readonly statusCode: number

  public constructor({ body, error, statusCode = HTTP_STATUS_CODE.ok }: RestResponseProps<Body>) {
    this._body = body
    this.error = error
    this.statusCode = statusCode
  }

  public get isFailure(): boolean {
    return !this.isSuccessful
  }

  public get isSuccessful(): boolean {
    return this.error === undefined && this.statusCode >= 200 && this.statusCode < 300
  }

  public get body(): Body {
    if (this.isFailure || this._body === undefined) {
      throw this.error ?? new AppError("Response body is unavailable", this.statusCode)
    }

    return this._body
  }

  public getBody(): Body {
    return this.body
  }

  public mapBody<Result>(mapper: (body: Body) => Result): RestResponse<Result> {
    return this.isSuccessful
      ? new RestResponse({ body: mapper(this.body), statusCode: this.statusCode })
      : new RestResponse({ error: this.error, statusCode: this.statusCode })
  }
}
